poa-token/src/contract.rs

```diff
@@ -56,6 +153,236 @@ impl Contract {
+        let near_sdk::PromiseResult::Successful(metadata_bytes) = env::promise_result(0) else {
+            env::panic_str(
+                "Setting metadata failed due to the promise failing at ft_metadata() call.",
+            )
+        };
+
+        let incoming_metadata = serde_json::from_slice::<FungibleTokenMetadata>(&metadata_bytes)
+            .unwrap_or_else(|e| {
+                env::panic_str(&format!(
+                    "Setting metadata failed due to unparsable promise data: {e}"
+                ))
+            });
+
+        let initial_storage_usage = env::storage_usage();
+
+        let to_set_metadata = FungibleTokenMetadata {
+            spec: FT_METADATA_SPEC.to_string(),
+            name: format!("Wrapped {}", incoming_metadata.name),
+            symbol: format!("w{}", incoming_metadata.symbol),
+            ..incoming_metadata
+        };
+
+        match self {
+            Contract::WrappableToken {
+                token: _,
+                metadata,
+                wrapped_token: _,
+            } => metadata.set(to_set_metadata),
+        }
+
+        self.metadata_mut().flush();
+
+        let end_storage_usage = env::storage_usage();
+
+        // Note that we use saturating sub here to prevent abuse. We do not refund Near for freed storage
+        let new_storage_byte_count = end_storage_usage.saturating_sub(initial_storage_usage);
+
+        let new_storage_cost = env::storage_byte_cost()
+            .checked_mul(u128::from(new_storage_byte_count))
+            .expect("Storage cost calculation overflow");
+
+        let storage_cost_refund = env::attached_deposit()
+            .checked_sub(new_storage_cost)
+            .unwrap_or_else(|| {
+                env::panic_str(&format!(
+                    "Not enough attached deposit for updating metadata: {}yNEAR < {}yNEAR",
+                    new_storage_cost.as_yoctonear(),
+                    env::attached_deposit().as_yoctonear()
+                ))
+            });
+
+        near_sdk::log!(
+            "Updated metadata with cost from size {}yNEAR to size {}yNEAR with cost {}yNEAR",
+            initial_storage_usage,
+            end_storage_usage,
+            new_storage_cost
+        );
+
+        Promise::new(caller_id).transfer(storage_cost_refund)
+    }
+
+    #[only(self, owner)]
+    #[payable]
+    pub fn set_wrapped_token_account_id(&mut self, token_account_id: AccountId) -> Promise {
+        assert_one_yocto();
+
+        require!(
+            self.wrapped_token().is_none(),
+            "Wrapped token is already set"
+        );
+
+        ext_ft_core::ext(token_account_id.clone())
+            .with_static_gas(BALANCE_OF_GAS)
+            .ft_balance_of(CURRENT_ACCOUNT_ID.clone())
+            .then(
+                Self::ext(CURRENT_ACCOUNT_ID.clone())
+                    .with_static_gas(DO_WRAP_TOKEN_GAS)
+                    .do_set_wrapped_token_account_id(token_account_id),
```

Comment:
What if the balance changes during this async promises? We need to "pause" the contract first..